[1.11] Add Smelting Recipes

diesieben07 opened the pull request:
Updated and reviewed version of #3141.

Experience is no longer handled by the smelting registry, the existing systems for this are untouched. This is because due to the way the furnace works it is not practical to determine the experience on a per-recipe basis.

diesieben07 committed:
Add furnace recipes

diesieben07 committed:
Update smelting recipes to 1.11.x

diesieben07 committed:
Update smelting recipes to 1.11.x

diesieben07 committed:
Add license header

diesieben07 commented:
There was a sorting mechanism mentioned by @LexManos, should that be similar in complexity to the existing RecipeSorter?

GotoLink commented on src/main/java/net/minecraftforge/common/smelting/SmeltingRecipeRegistry.java:
```diff
@@ -0,0 +1,194 @@
+import com.google.common.reflect.Reflection;
+
+/**
+ * Recipe for smelting recipes, should be used over the GameRegistry methods and vanilla's FurnaceRecipes class.
```
"Registry" for...

diesieben07 replied:
Fixed.

GotoLink commented on src/main/java/net/minecraftforge/common/smelting/SmeltingRecipeRegistry.java:
```diff
@@ -0,0 +1,194 @@
+    /**
+     * Default time a recipe takes to cook.
+     */
+    public static final int DEFAULT_COOK_TIME = 200;
```
Maybe add in the comment this value is extracted from TileEntityFurnace ?

diesieben07 replied:
And fixed.

mezz commented on src/main/java/net/minecraftforge/fml/common/registry/GameRegistry.java:
```diff
@@ -242,16 +242,19 @@ public static void addRecipe(IRecipe recipe)
     }
 
-    public static void addSmelting(Block input, @Nonnull ItemStack output, float xp)
+    @Deprecated // Use FurnaceRecipeRegistry
```
You should add a javadoc that links to the new method.

mezz replied:
Also it looks like you removed a `@Nonnull` on the ItemStack parameter accidentally.

mezz commented on src/main/java/net/minecraftforge/fml/common/registry/GameRegistry.java:
```diff
@@ -242,16 +242,19 @@ public static void addRecipe(IRecipe recipe)
         FurnaceRecipes.instance().addSmeltingRecipeForBlock(input, output, xp);
     }
 
+    @Deprecated // Use FurnaceRecipeRegistry
```
You should add a javadoc that links to the new method.

mezz commented on src/main/java/net/minecraftforge/fml/common/registry/GameRegistry.java:
```diff
@@ -242,16 +242,19 @@ public static void addRecipe(IRecipe recipe)
         FurnaceRecipes.instance().addSmelting(input, output, xp);
     }
 
+    @Deprecated // Use FurnaceRecipeRegistry
```
You should add a javadoc that links to the new method.

mezz commented on src/main/java/net/minecraftforge/common/smelting/SmeltingRecipeRegistry.java:
```diff
@@ -0,0 +1,194 @@
+    }
+
+    @Deprecated // internal API
+    public static void getAllPossibleItems(Collection<Item> target)
```
`getAllPossibleOutputs` may be a better name. 
Probably no need to mark this as an internal API, I'm sure someone could find a use for it.

mezz commented on src/main/java/net/minecraftforge/common/smelting/VanillaSmeltingRecipe.java:
```diff
@@ -0,0 +1,66 @@
+import net.minecraft.item.ItemStack;
+import net.minecraftforge.oredict.OreDictionary;
+
+final class VanillaSmeltingRecipe implements SmeltingRecipe
```
Why is this different from `SimpleSmeltingRecipe`? I think you should try to handle vanilla recipes without a special case.

mezz commented on src/main/java/net/minecraftforge/common/smelting/SmeltingRecipeRegistry.java:
```diff
@@ -0,0 +1,194 @@
+        return 0;
+    }
+
+    @Deprecated // internal API, do not use!
```
I don't like using `@Deprecated` in this way. If someone wants to add a vanilla-style recipe, they should be able to.

diesieben07 committed:
Javadoc improvements: fix typo & clarify default smelting time

mezz commented on src/main/java/net/minecraftforge/common/smelting/SmeltingRecipeRegistry.java:
```diff
@@ -0,0 +1,194 @@
+     *
+     * @param recipe the recipe
+     */
+    public static void addRecipe(@Nonnull SmeltingRecipe recipe)
```
All these `addRecipe` methods should support an XP parameter that gets used somehow, otherwise these are not a direct replacement of the `GameResgitry.addSmelting` methods because there is no support for recipes with XP.

mezz reviewed (changes requested):
Looks good but there are some things to address before it is ready.

diesieben07 committed:
One more Javadoc fix

diesieben07 committed:
- Add proper support for experience handling instead of just saying "not here!"
- Get rid of VanillaSmeltingRecipe and just use the normal methods
- Expose getAllPossibleOutputs as an actual API method
- Add Javadocs to the GameRegistry methods directing to the new place

diesieben07 committed:
- Add proper support for experience handling instead of just saying "not here!"
- Get rid of VanillaSmeltingRecipe and just use the normal methods
- Expose getAllPossibleOutputs as an actual API method
- Add Javadocs to the GameRegistry methods directing to the new place

diesieben07 commented:
I addressed all your remarks @mezz.

diesieben07 committed:
two more missing license headers

diesieben07 committed:
re-add missing @NonNull

diesieben07 committed:
make the method order a bit more consistent

mezz commented:
Thanks, looks pretty good.

I think you should avoid efficient ItemStack comparison and just do it the same way vanilla does. There is too much potential for bugs if you diverge from their implementation. For instance, the hashing thing does not account for wildcard meta like `FurnaceRecipes.compareItemStacks` does.  

I think adding a method like `setSmeltingExperience(ItemStack stack, float experience)` to `FurnaceRecipes` would be safer.

diesieben07 commented:
>  For instance, the hashing thing does not account for wildcard meta like FurnaceRecipes.compareItemStacks does.

Yes indeed it does not, but getExperience in SmeltingRecipeRegistry does. The hashing is only an implementation detail. But if you think patching FurnaceRecipes and using that as a backend is the way to go...

diesieben07 committed:
- Use existing vanilla mechanics for experience
- Add getters for recipe data

diesieben07 commented:
Updated as discussed on IRC.

diesieben07 committed:
Add I prefix to interfaces and remove Item::getSmeltingExperience

mezz commented:
As discussed in IRC, this is now a Breaking Change because it removes `Item.getSmeltingExperience`.
It could be deprecated instead but we might as well get rid of it now while we can in 1.11.

diesieben07 committed:
Fix javadocs and formatting errors

mezz commented on patches/minecraft/net/minecraft/item/crafting/FurnaceRecipes.java.patch:
```diff
@@ -1,20 +1,45 @@
-+
-         for (Entry<ItemStack, Float> entry : this.field_77605_c.entrySet())
++        if (true) return net.minecraftforge.common.smelting.SmeltingRecipeRegistry.getOutput(p_151395_1_);
+         for (Entry<ItemStack, ItemStack> entry : this.field_77604_b.entrySet())
```
Just making a note that the patch above *looks* insane but it's actually fine.

mezz commented:
Tested and it looks good to me.

ghost commented:
Hey guys . I was recommend I join this discussion . 

I would like the ability to conditionally control Furnace recipes turn them on or off (that is some furnaces can do this and other cannot)

for ex  specific furnaces may do 
off Raw Meat -> Cooked Meat (normally is this on but is now off)
on Raw Meat -> Steamed Meat (normally this is off but is now on)

or
on Pretty Meat -> Nuggets (normally off but i also turn this on)

so furnace all the sudden can do 2 things that other furnaces can't. Not sure how would this work behind the scenes or if is even possible. I was told event hooks could do it. I was even thinking having a custom smelting recipes that be associated with custom furnaces and added or removed dynamically

mezz commented:
That sounds very confusing from a player's point of view.
Why not make your own furnace that builds on top of the vanilla one? You can add all the functionality you want there.

ghost commented:
@mezz  Well I guess it is indeed. I guess what I need not everyone else needs. Maybe simply adding extra hooks such as smeltingItemEvent for example that fires right before a furnace is about to smelt something might help me out a lot

mezz commented:
@trollworkout Ok. I think that's different from this PR though so let's keep any further discussion in the original issue you opened.

ghost commented:
@mezz  ok! Understood

mezz commented:
This needs an update to resolve a merge conflict.

CrafterKina commented:
Could you add features like #1693?
I'm sorry if this is an unrelated issue.

diesieben07 commented:
Rebased onto current 1.11.x.

mezz commented:
@CrafterKina I don't think we need that much control over the furnace. 
It may be confusing to players who expect the furnace to work normally. 
If you want very custom behavior, you can make your own furnace. That is what many mods do.

CrafterKina commented:
@mezz The function I proposed is the same as the behavior of sponge in vanilla, so I do not think it leads to confuse the player.

diesieben07 committed:
rebase onto 1.11.x again

diesieben07 commented:
Rebased and resolved conflicts again.

mezz commented:
It looks like this is probably not going to be pulled. Stepping back to look at it, this really is a lot to add just to get custom burn times and a small amount of extra flexibility.
Thanks for bearing with me through the review process, I think we all at least gained something from doing that.

diesieben07 commented:
Not ever? Well... that was a bunch of wasted time then...

Also, just as a reminder, this _was_ originally just a simple hook for cook time. All this recipe stuff was added because Lex suggested it: https://github.com/MinecraftForge/MinecraftForge/pull/3141#issuecomment-239571359

And now this gets closed because "it does all this recipe stuff just for cook time". What?

mezz commented:
That's my guess. 1.11 has a recommended build and this was still not pulled. Waiting on a response from Lex.

KnightMiner commented:
Also, it allows more then just cook times, it allows furnace recipes to have more complicated matching, such as oredictionary, or the thing I was interested in: NBT.

ghost commented:
Can we get this please. Would be highly appreciated. Thank you kindly in advance.
